File: dashboard/src/stores/webhooks.ts
import { create } from "zustand";
import { DeckApiError, fetchApi } from "@/lib/errors";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface Webhook {
  id: string;
  name: string;
  url: string;
  secret: string | null;
  events: string[];
  enabled: boolean;
  consecutiveFailures: number;
  lastFiredAt: string | null;
  lastStatus: number | null;
  createdAt: string;
  updatedAt: string;
}

export interface WebhookDelivery {
  id: string;
  webhookId: string;
  eventType: string;
  statusCode: number | null;
  error: string | null;
  durationMs: number | null;
  success: boolean;
  isRetry: boolean;
  createdAt: string;
}

export interface CreateWebhookInput {
  name: string;
  url: string;
  secret?: string;
  events: string[];
  enabled?: boolean;
}

export type UpdateWebhookInput = Partial<CreateWebhookInput>;

// ---------------------------------------------------------------------------
// Store
// ---------------------------------------------------------------------------

interface WebhookState {
  webhooks: Webhook[];
  selectedWebhookId: string | null;
  deliveries: WebhookDelivery[];
  loading: boolean;
  error: string | null;

  selectWebhook: (id: string | null) => void;
  fetchWebhooks: () => Promise<void>;
  createWebhook: (input: CreateWebhookInput) => Promise<void>;
  updateWebhook: (id: string, input: UpdateWebhookInput) => Promise<void>;
  deleteWebhook: (id: string) => Promise<void>;
  fetchDeliveries: (webhookId: string) => Promise<void>;
  testWebhook: (webhookId: string) => Promise<void>;
}

export const useWebhookStore = create<WebhookState>((set, get) => ({
  webhooks: [],
  selectedWebhookId: null,
  deliveries: [],
  loading: false,
  error: null,

  selectWebhook: (id) => set({ selectedWebhookId: id, deliveries: [] }),

  fetchWebhooks: async () => {
    set({ loading: true, error: null });
    try {
      const data = await fetchApi<{ webhooks: Webhook[] }>("/api/webhooks");
      set({ webhooks: data.webhooks, loading: false });
    } catch (err) {
      set({
        error: err instanceof DeckApiError ? err.body.error : "Failed to fetch webhooks",
        loading: false,
      });
    }
  },

  createWebhook: async (input) => {
    set({ error: null });
    try {
      await fetchApi<Webhook>("/api/webhooks", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(input),
      });
      await get().fetchWebhooks();
    } catch (err) {
      set({ error: err instanceof DeckApiError ? err.body.error : "Failed to create webhook" });
    }
  },

  updateWebhook: async (id, input) => {
    set({ error: null });
    try {
      await fetchApi<Webhook>(`/api/webhooks/${id}`, {
        method: "PATCH",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(input),
      });
      await get().fetchWebhooks();
    } catch (err) {
      set({ error: err instanceof DeckApiError ? err.body.error : "Failed to update webhook" });
    }
  },

  deleteWebhook: async (id) => {
    set({ error: null });
    try {
      await fetchApi<{ deleted: boolean }>(`/api/webhooks/${id}`, { method: "DELETE" });
      const { selectedWebhookId } = get();
      set({ selectedWebhookId: selectedWebhookId === id ? null : selectedWebhookId });
      await get().fetchWebhooks();
    } catch (err) {
      set({ error: err instanceof DeckApiError ? err.body.error : "Failed to delete webhook" });
    }
  },

  fetchDeliveries: async (webhookId) => {
    try {
      const data = await fetchApi<{ deliveries: WebhookDelivery[] }>(
        `/api/webhooks/${webhookId}/deliveries`,
      );
      set({ deliveries: data.deliveries });
    } catch {
      set({ deliveries: [] });
    }
  },

  testWebhook: async (webhookId) => {
    set({ error: null });
    try {
      await fetchApi<{
        success: boolean;
        statusCode: number | null;
        durationMs: number | null;
        error: string | null;
        deliveryId: string;
      }>(`/api/webhooks/${webhookId}/test`, { method: "POST" });
      // Refresh deliveries to show the test result
      await get().fetchDeliveries(webhookId);
    } catch (err) {
      set({ error: err instanceof DeckApiError ? err.body.error : "Test delivery failed" });
    }
  },
}));

File: deck-go/backend/internal/localstore/webhooks.go
package localstore

type Webhook struct {
	ID                  string   `json:"id"`
	Name                string   `json:"name"`
	URL                 string   `json:"url"`
	Secret              *string  `json:"secret"`
	Events              []string `json:"events"`
	Enabled             bool     `json:"enabled"`
	ConsecutiveFailures float64  `json:"consecutiveFailures"`
	LastFiredAt         *string  `json:"lastFiredAt"`
	LastStatus          *float64 `json:"lastStatus"`
	CreatedAt           string   `json:"createdAt"`
	UpdatedAt           string   `json:"updatedAt"`
}

type WebhookDelivery struct {
	ID             string   `json:"id"`
	WebhookID      string   `json:"webhookId"`
	EventType      string   `json:"eventType"`
	Payload        string   `json:"payload"`
	StatusCode     *float64 `json:"statusCode"`
	ResponseBody   *string  `json:"responseBody"`
	Error          *string  `json:"error"`
	DurationMs     float64  `json:"durationMs"`
	Attempt        float64  `json:"attempt"`
	IsRetry        bool     `json:"isRetry"`
	ParentDelivery *string  `json:"parentDeliveryId"`
	Success        bool     `json:"success"`
	NextRetryAt    *float64 `json:"nextRetryAt"`
	CreatedAt      string   `json:"createdAt"`
}

func GetWebhookStore() *SliceStore[Webhook] {
	return NewSliceStore[Webhook]("webhooks.json")
}

func GetWebhookDeliveryStore() *SliceStore[WebhookDelivery] {
	return NewSliceStore[WebhookDelivery]("webhook-deliveries.json")
}
